### PR TITLE
1799: Optimise data fetching for indicators

### PR DESCRIPTION
**Pull request by kael89**
### Issue #:

https://github.com/beyondessential/tupaia-backlog/issues/1799

Code here is already reviewed in other PRs:
* https://github.com/beyondessential/tupaia/pull/1850
* https://github.com/beyondessential/tupaia/pull/1852
* https://github.com/beyondessential/tupaia/pull/1873
* https://github.com/beyondessential/tupaia/pull/1919
* https://github.com/beyondessential/tupaia/pull/2071
* https://github.com/beyondessential/tupaia/pull/2087
* https://github.com/beyondessential/tupaia/pull/2100

### CODE REVIEW

**Comment by kael89** [comment]
@edmofro this is the master feature branch where everything is merged. Please hold off review on this one - I have to merge `dev` and resolve potential conflicts. Will let you know when it's ready

**Comment by kael89** [comment]
@edmofro I merged dev and resolved hard/soft conflicts... Ready for re-review!

File: packages/indicators/src/__tests__/__integration/fixtures/arithmeticAnalyticsFixtures.ts [hunk 1]
```diff
@@ -32,7 +32,7 @@ const ARRAY_SURVEY_RESPONSES: ArraySurveyResponse[] = [
   ['Births', 'GR', '2019-01-01', { Female: '1', Male: '2' }],
   ['Births', 'GR', '2020-11-30', { Female: '3', Male: '4' }],
   ['Births', 'GR', '2020-12-29', { Female: '60', Male: '80' }],
-  ['Births', 'GR', '2020-12-30', { Female: '6', Male: '8' }],
+  ['Births', 'GR', '2020-12-31', { Female: '6', Male: '8' }],
```
Comment (kael89) [inline packages/indicators/src/__tests__/__integration/fixtures/arithmeticAnalyticsFixtures.ts:35]: The existing value (`30`) would also work... `30` was used to enable a match with the previous month (`2020-11-30`), but after https://github.com/beyondessential/tupaia/pull/1941 any month day can be used (which is the desired behaviour)

File: packages/indicators/src/Builder/ArithmeticBuilder/ArithmeticBuilder.ts [hunk 2]
```diff
@@ -80,17 +82,31 @@ export class ArithmeticBuilder extends Builder {
...
+  private fetchParameterAnalytics = async (fetchOptions: FetchOptions) => {
+    const analytics = await this.api.buildAnalyticsForBuilders(this.paramBuilders, fetchOptions);
+    const analyticsByElement = groupBy(analytics, 'dataElement');
+
+    return Object.entries(analyticsByElement)
+      .map(([element, analyticsForElement]) => {
+        const aggregationList = this.config.aggregation[element];
+        return this.api
+          .getAggregator()
+          .aggregateAnalytics(analyticsForElement, aggregationList, fetchOptions.period);
```
Comment (kael89) [inline packages/indicators/src/Builder/ArithmeticBuilder/ArithmeticBuilder.ts:107]: This is decoupled fetching/aggregation of analytics in practice. It's not required by any existing use case (it isn't supported in `dev`, after all), but it's required by one new [indicator](https://github.com/beyondessential/tupaia/blob/1779-optimise-data-fetching-for-indicators/packages/indicators/src/__tests__/__integration/fixtures/arithmeticAnalyticsFixtures.ts#L122) in the integration tests

File: packages/indicators/src/__tests__/__integration/fixtures/arithmeticAnalyticsFixtures.ts (hunk 1 above)
Comment (edmofro) [reply packages/indicators/src/__tests__/__integration/fixtures/arithmeticAnalyticsFixtures.ts:35]: A nice improvement :-)

File: packages/indicators/src/Builder/ArithmeticBuilder/ArithmeticBuilder.ts (hunk 2 above)
Comment (edmofro) [reply packages/indicators/src/Builder/ArithmeticBuilder/ArithmeticBuilder.ts:107]: Yeah nice 👍

**Review by edmofro — APPROVED** [review]
Just reviewed the final two commits, assuming all of the previous have already been reviewed in separate PRs (correct me if I'm wrong)

LGTM!

**Comment by kael89** [comment]
We discussed with @chris-bes that we can test the changes here together with his [Count indicator](https://github.com/beyondessential/tupaia/pull/2114) code.

I changed the base branch for this to an epic which will include code for the two cards ([#1799](https://github.com/beyondessential/tupaia-backlog/issues/1799), [#1378](https://github.com/beyondessential/tupaia-backlog/issues/1378)), and will merge the code there to get it ready for testing

**Pull request merged**